[MRG before #12069] KernelPCA: raise Errors and Warnings according to eigenvalue decomposition numerical/conditioning issues

smarie opened the pull request:
This PR fixes #12140 by performing some numerical and conditioning checks on the eigenvalues found from kernel decomposition.

This PR is dependent on https://github.com/scikit-learn/scikit-learn/pull/12143, that solves an issue happening in `transform` when zero eigenvalues are present in the kernel.

smarie committed:
Minor edits for clarity:
 - Added a few comments to clarify `_fit_transform`, `fit_transform` and `transform`.
 - Uniformized the numpy coding style for `fit` and `fit_transform`.

smarie committed:
Fixed #12141.

smarie committed:
Added exceptions and warnings in case of numerical issues and bad kernel conditioning. Fixes #12140

smarie commented:
Needs review: the only failing check is coverage. To improve it we would need to produce tests where the kernel matrix has artificial defects. Would it be ok ? How to check that warnings are properly raised ?

smarie committed:
Added a `check_kernel_eigenvalues` validation method (with doctests). It is used in `KernelPCA`. Added corresponding test `test_errors_and_warnings` for KernelPCA.

smarie commented:
While waiting for the review I proposed a way to fix the coverage and improve maintainability: the method to check the kernel eigenvalues (`check_kernel_eigenvalues`) is now completely independent, in the `validation` submodule. It comes with some doctests.

In addition I created a test for kPCA to check that warnings and errors are raised correctly in case of bad conditioning.

smarie committed:
Added an additional check in the test, in order to understand the strange failure in Travis.

smarie committed:
Fixed test_errors_and_warnings: a wrong solver name had been introduced ('randomized' is not available in this branch !)

smarie committed:
kPCA test_errors_and_warnings: Added a few extra checks in the test to always check the inner check_kernel_eigenvalues method before the kPCA fit.

smarie committed:
Fixed test `test_errors_and_warnings`. Our identity centerer was replaced during call to fit() - we now execute the test directly on `_fit_transform` instead, so that the matrix is untouched.

smarie committed:
Fixed doctest in `check_kernel_eigenvalues`

smarie committed:
Final doctest fixes for `check_kernel_eigenvalues` (locally validated this time :) ).

smarie committed:
pytest doc fix: added normalize whitespace everywhere since there were still errors of this kind.

smarie committed:
Added test corresponding to issue #12141

smarie committed:
Following code review from @NicolasHug : now only scaling eigenvectors where there is no zero division, to avoid numpy warnings.

smarie committed:
Following code review from @NicolasHug : now checking that there is no zero division numpy warning.

smarie committed:
Following code review from @NicolasHug : now using pytest warning capture to perform fine-grain warning assertion

smarie committed:
Merge branch 'master' into kPCA_fix_issue_12141

smarie committed:
Merge was wrong - removing useless import

smarie committed:
Merge branch 'kPCA_fix_issue_12141' into kPCA_fix_issue_12140

# Conflicts:
#	sklearn/decomposition/tests/test_kernel_pca.py

smarie commented:
I had some afterthoughts on one of the warnings: indeed it is normal to find quasi-zero eigenvalues when the number of samples is high enough (my intuition would be that it is when this number is larger than the underlying distribution's manifold dimensionality, but maybe in [this paper](https://homepages.inf.ed.ac.uk/ckiw/postscript/gram.pdf) there are better explanations).

For this reason I will push a new commit where there is no warning by default about zero eigenvalues.

smarie committed:
Afterthoughts on the need to warn when gram matrix eigenvalues are small: this is most probably a common case especially when the number of samples gets high. Removing the warning by default.

smarie committed:
Fixed test: no warning is raised in presence of small eigenvalues anymore, this is normal behaviour as the number of samples increases.

smarie commented:
All set ! Ready to merge.

jnothman commented on sklearn/decomposition/kernel_pca.py:
```diff
@@ -8,7 +8,8 @@
 
 from ..utils import check_random_state
-from ..utils.validation import check_is_fitted, check_array
+from ..utils.validation import check_is_fitted, check_array, \
```
please use parentheses rather than `\` for line continuation

smarie replied:
ok

jnothman commented on sklearn/decomposition/kernel_pca.py:
```diff
@@ -295,8 +323,16 @@ def transform(self, X):
+
+        # scale eigenvectors (properly account for null-space for dot product)
+        nz = np.flatnonzero(self.lambdas_)
+        scaled_alphas = np.zeros(self.alphas_.shape)
```
to be sure of dtype matching, use zeros_like.

smarie replied:
ok. I'll push this to #12143, merge back and push this

jnothman commented on sklearn/decomposition/kernel_pca.py:
```diff
@@ -295,8 +323,16 @@ def transform(self, X):
+        # scale eigenvectors (properly account for null-space for dot product)
+        nz = np.flatnonzero(self.lambdas_)
+        scaled_alphas = np.zeros(self.alphas_.shape)
+        scaled_alphas[:, nz] = self.alphas_[:, nz] / np.sqrt(self.lambdas_[nz])
```
Is this a separate bug fix to the rest of the PR?

smarie replied:
yes this PR is the second of a series of 3. that particular bit of code comes from https://github.com/scikit-learn/scikit-learn/pull/12143 .

jnothman reviewed (commented):
Not able to review this properly now, but someone might be keen if you resolve conflicts with master, thanks @smarie 

Are there parts of the KernelPCA documentation that deserve updatating?
It might also be a good idea to add an entry to doc/whats_new/v0.21.rst to clearly summarise what the effect of this fix and change is on users (helping the reviewer).

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into kPCA_fix_issue_12141

Commit:
Improved zero array initialization according to code review https://github.com/scikit-learn/scikit-learn/pull/12145/files/2112f5d0278e6342d93e9649d3b0da6021d30f98#r261052208

Commit:
Merge branch 'kPCA_fix_issue_12141' of https://github.com/smarie/scikit-learn into kPCA_fix_issue_12140

# Conflicts:
#	sklearn/decomposition/kernel_pca.py
#	sklearn/decomposition/tests/test_kernel_pca.py
#	sklearn/utils/validation.py

Commit:
Merge branch 'kPCA_fix_issue_12141' of https://github.com/smarie/scikit-learn into kPCA_fix_issue_12140

Commit:
Improved import style according to code review https://github.com/scikit-learn/scikit-learn/pull/12145#discussion_r261052208

Commit:
Updated what's new for this PR

Commit:
Merge branch 'kPCA_fix_issue_12141' of https://github.com/smarie/scikit-learn into kPCA_fix_issue_12140

Commit:
Updated what's new for this PR

smarie commented:
@jnothman I updated what's new following your advice. The documentation does not need to be updated, this is ready to Merge (this PR is the second of a series of 3: #12143, #12145 and #12069)

NicolasHug commented on sklearn/utils/validation.py:
```diff
@@ -982,3 +982,121 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
         raise ValueError('`{}`= {}, must be <= {}.'.format(name, x, max_val))
+
+
+def check_kernel_eigenvalues(lambdas, warn_on_zeros=False):
```
`utils.validation.py` is used for input checking so I don't think this belongs here. It should probably be in `kernel_pca.py` instead.

Also it should be private: users shouhldn't be allowed to use it directly (nor should they need to).

smarie replied:
I see. Note that `check_kernel_eigenvalues` is something that could be used everywhere where a kernel spectrum is computed, whether using exact or approximation methods. Or more generally a semi positive definite matrix's spectrum. So I thought that I would put it somewhere more central so that other developers in sklearn get a chance to try to reuse it. I thought that `utils.validation.py` was a good place but apparently I was wrong, where would be a better place for common internal validation tools? Thanks !

jnothman replied:
I don't mind this here so much...

NicolasHug replied:
Ok, then we might want to update the current module's docstring which says 
`"""Utilities for input validation"""`.

@smarie, since the function is relevant for any semi positive definite matrix, maybe change the name to e.g. (sorry, long name) `check_semi_positive_definite_eigenvalues()`

Also, would it be relevant to use it in the `PCA` code as well?

smarie replied:
If size matters, `check_spd_eigenvalues()` is almost as understandable.
If we go in that direction, I'll have to rename `KernelWarning` into `SpdSpectrumWarning`.

It does indeed apply to `PCA` and in general to all places where we use `eig` (non-symmetric matrix) or `eigh` (symmetric matrix) to compute the spectrum of a SPD matrix. I would be surprised that we use `eig` anywhere, but just in case...

smarie replied:
note: `SPDSpectrumWarning` would look better than `SpdSpectrumWarning`

thomasjpfan committed:
Update doc/whats_new/v0.21.rst

Co-Authored-By: smarie <sylvain.marie@schneider-electric.com>

Commit:
grammar fix

Commit:
Referenced the issue

Commit:
Grammar fix

Commit:
Issue number replaced with PR number

Commit:
renamed `nz` `nonzeros` following review

Commit:
Fixed 80 characters max issue + improved comment as per review

Commit:
Fixed comment as per review

jnothman commented on sklearn/utils/validation.py:
```diff
@@ -982,3 +982,121 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+     If this check fails and `warn_on_zeros=True`, it raises a KernelWarning.
+     All the eigenvalues that are too small are then set to zero.
+
+    Note: the returned array is converted to numpy array.
```
Is it possible that if a numpy array is passed in it will be modified in-place?

smarie replied:
Good catch. I thought that `lambdas = np.real(lambdas)` was creating a copy but that's not the case.

I now use `lambdas = np.real(np.copy(lambdas))` in line 1070

jnothman commented on sklearn/utils/validation.py:
```diff
@@ -982,3 +982,121 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+
+def check_kernel_eigenvalues(lambdas, warn_on_zeros=False):
+    """Checks that the provided array of kernel eigenvalues for numerical or
```
Checks kernel eigenvalues for conditioning issues

is sufficient. The rest can be expanded on below

smarie replied:
Ok.

jnothman commented on sklearn/utils/validation.py:
```diff
@@ -982,3 +982,121 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+    It checks:
+
+     - that there are no significant imaginary parts in eigenvalues (more than
```
Please use hanging indentation for restructured text lists

smarie replied:
I think I got what you mean: I remove the space before the dash and indent the text within each bullet as follows:

```python
    """Checks kernel eigenvalues for conditioning issues.

    Checks the provided array of kernel eigenvalues for numerical or
    conditioning issues and returns a fixed validated version.

    It checks:

    - that there are no significant imaginary parts in eigenvalues (more than
      1e-5 times the maximum real part). If this check fails, it raises a
      ValueError. Otherwise all non-significant imaginary parts that may remain
      are removed.

    - that eigenvalues are not all negative. If this check fails, it raises a
      ValueError

    ...
```

jnothman commented on sklearn/utils/validation.py:
```diff
@@ -982,3 +982,121 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+    Note: the returned array is converted to numpy array.
+
+    >>> check_kernel_eigenvalues((1, 2))      # nominal case
```
Please follow numpy: put this under an Examples heading below the Returns section

smarie replied:
ok

jnothman reviewed (commented):
Just looking at the helper, which deserves unit tests of its own

jnothman commented:
Are you saying that we should not call this "*input* validation"? Yes,
maybe you're right it's a bit out of place here. Not sure where is
better, though.

NicolasHug commented:
Yes, this is not strictly speaking *input* validation. OTOH it seems that this helper could be useful in multiple places. So far I found these (looking for calls to `eigh`):

- robust_covariance.py, see https://github.com/scikit-learn/scikit-learn/pull/8328 which could leverage this now
- ridge.py:  ` v, Q = linalg.eigh(K)`
- spectral_embedding_.py:   `lambdas, diffusion_map = eigh(laplacian)`
- fastica_.py:    ` s, u = linalg.eigh(np.dot(W, W.T))`
- discriminant_analysis.py: `linalg.eigh(Sb, Sw)` (not sure if `Sb = St - Sw` is SPD though)

Given the breadth of potential uses of the helper and since there's no other convention for such checkers yet, then I agree it's OK to keep it in `utils/validation.py`

Commit:
Fixed comment as per review

Commit:
Fixed test as per review

Commit:
Merge branch 'kPCA_fix_issue_12141' into kPCA_fix_issue_12140

NicolasHug commented on sklearn/utils/validation.py:
```diff
@@ -982,3 +982,121 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+     If this check fails and `warn_on_zeros=True`, it raises a KernelWarning.
+     All the eigenvalues that are too small are then set to zero.
+
+    Note: the returned array is converted to numpy array.
```
Maybe just remove this line since it's redundant with the `Returns` section.

smarie replied:
ok

Commit:
Fixed `check_kernel_eigenvalues` to return a copy (as per code review)

Commit:
Fixed `check_kernel_eigenvalues` docstring as per code review

Commit:
Added name in authors list

Commit:
Fixed flake8 error

smarie commented:
Note that in all the cases you mention, the spd matrix often depends on user input, and sometimes quite directly (when a custom kernel is provided). So indeed it does not sound like a bad choice... for now

smarie commented:
So the last remaining point here is about the name/scope of the checker (see unresolved conversation above).

NicolasHug commented:
@smarie we also need to wait for the previous issue to be merge so that the diff here looks smaller

NicolasHug commented:
@smarie can you please merge master now that #12143 has been merged?

Commit:
Merge remote-tracking branch 'original/master' into kPCA_fix_issue_12140

smarie commented:
I did, but it does not seem to change the diff view.. is there something I need to do ?

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into kPCA_fix_issue_12140

smarie commented on doc/whats_new/v0.21.rst:
```diff
@@ -89,6 +89,14 @@ Support for Python 3.4 and below has been officially dropped.
+  consistency of results across solvers (different choices for `eigen_solver`),
+  including approximate solvers such as `'randomized'` and `'lobpcg'` (see
+  :issue:`12068`).
+  :issue:`12140` by :user:`Sylvain Marié <smarie>`
```
I'll replace with the PR number

Commit:
Fixed issue number > PR number

NicolasHug commented on sklearn/decomposition/kernel_pca.py:
```diff
@@ -211,6 +212,10 @@ def _fit_transform(self, K):
+        # make sure that there are no numerical or conditioning issues
+        self.lambdas_ = check_kernel_eigenvalues(self.lambdas_)
+
+
```
remove 1 blank line

smarie replied:
ok

NicolasHug commented on sklearn/decomposition/tests/test_kernel_pca.py:
```diff
@@ -272,3 +276,100 @@ def test_nested_circles():
+    K = [[5, 0],
+         [0, -6e-5]]
+    # check that the inner method works
+    assert_warns(KernelWarning,
```
instead of `assert_warns` please use `with pytest.warns(KernelWarning, match=warning_mesg):` instead

smarie replied:
ok

NicolasHug commented on sklearn/decomposition/tests/test_kernel_pca.py:
```diff
@@ -272,3 +276,100 @@ def test_nested_circles():
+        kpca._centerer = IdentityKernelTransformer()
+        K = kpca._get_kernel(K)
+        # note: we can not test 'fit' because _centerer would be replaced
+        assert_no_warnings(lambda: kpca._fit_transform(K))
```
Please use `with pytest.warns(None) as w: ... assert not w`

smarie replied:
ok.

NicolasHug commented on sklearn/exceptions.py:
```diff
@@ -154,3 +155,10 @@ class UndefinedMetricWarning(UserWarning):
+
+
+class KernelWarning(UserWarning):
+    """Custom warning to capture kernel issues
```
It's not clear what a kernel is in this context

smarie replied:
Yes, I'll use `SPDSpectrumWarning`.

NicolasHug reviewed (commented):
More reviews. Sorry, nothing very substantial yet but we'll get there!

About naming, I like `check_spd_eigenvalues` and `SPDSpectrumWarning`.

I'm not a fan of naming things with *Kernel* because it can refer to multiple things (kernel trick, null space...). In particular it is not obvious in the docs what `kernel eigenvalues` are.

Commit:
Removed 1 blank line

Commit:
As per code review: since `assert_warns` and `assert_no_warnings` are to be rolled out, replaced them with `pytest.warns`.

Commit:
As per code review: renamed `check_kernel_eigenvalues` into `check_psd_eigenvalues` and `KernelWarning` into `PSDSpectrumWarning`

smarie commented:
Done. Note that I found out that the correct term is `PSD` (Positive semidefinite), so that's the naming I adopted.

Commit:
Minor edit: line too long

NicolasHug commented:
Thansk @smarie 	

As Joel noted before, the `check_psd_eigenvalues` util would need a test on its own, in `test_validation.py`. The tests for KernelPCA could be lighter then

NicolasHug commented:
Basically everything you have in the example sections could be a test (they're already doctests but it's better to have them in `test_validation.py`)

Commit:
Added tests for `check_psd_eigenvalues`

Commit:
Fixed warning message for bad conditioning: now discarding exact zero eigenvalues.

smarie commented:
Done. I left the doctests though, let me know if I need to remove them entirely.

NicolasHug commented on sklearn/decomposition/tests/test_kernel_pca.py:
```diff
@@ -272,3 +275,104 @@ def test_nested_circles():
     assert_equal(train_score, 1.0)
+
+
+def test_errors_and_warnings():
```
Call this `test_kernel_conditioning` maybe?

Right now the tests don't feel very *unitary*, since you rely on the fact that the `_centerer` attribute isn't overridden when calling `fit_transform`.
Since we have a dedicated test for `check_spd_values` now, I think it'd be enough here to simply make sure that `check_spd_values` gets called, e.g. by passing a pathological X and making sure we get a warning / error from `check_spd_values`.

smarie replied:
If I remember well my problem was that could not create a pathological `X` here without the subsequent PR (using the approximate solver). I'll try again..

smarie replied:
Ok I gave up handcrafting a kernel as I do not master enough the spectrum of centered kernel matrices. So back to you idea of pathological X, I found an example leading to a small non-zero eigenvalue, I simply check that it is correctly set to zero by `check_psd_eigenvalues`.

NicolasHug commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -835,3 +836,63 @@ def test_check_scalar_invalid(x, target_name, target_type, min_val, max_val,
+    when warn_on_zeros is set to True, and does not when it is set to False"""
+
+    input = (5, 4e-12)
+    output = np.array([5., 0.])
```
Here and everywhere else, you shouldn't need to explicitly set floats (`5.` -> `5`)

smarie replied:
ok

NicolasHug reviewed (commented):
Some comments and nitpicks, also reviewed the tests. This looks good overall!

How did you choose the thresholds @smarie ? Are they somehow arbitrary? Would it make sense to pass them as parameters (e.g. for the float issue you're mentioning)?

Commit:
Added `check_psd_eigenvalues` to `classes.rst` for auto documentation

smarie commented on sklearn/utils/__init__.py:
```diff
@@ -57,7 +57,7 @@ class Parallel(_joblib.Parallel):
 __all__ = ["murmurhash3_32", "as_float_array",
            "assert_all_finite", "check_array",
-           "check_random_state",
+           "check_random_state", "check_psd_eigenvalues",
```
Ok. I updated `classes.rst` for now because I'm more in favour of letting users play with it and possibly suggest to improve it further ; but we can surely wait for a third point of view here.

smarie replied:
(done - this is private)

smarie commented on sklearn/utils/validation.py:
```diff
@@ -966,3 +967,134 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+
+def check_psd_eigenvalues(lambdas, warn_on_zeros=False):
+    """Checks the eigenvalues of a positive semidefinite (PSD) matrix.
```
I did not know that one :) that's not the way it is done in other languages, but of course I'll adapt and remove the 's' in 'checks'.

Commit:
Fixed one-liner docstring as per pep257.

Commit:
Using double barticks in rst docstring as per review

smarie commented on sklearn/utils/validation.py:
```diff
@@ -966,3 +967,134 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+    - that there are no significant imaginary parts in eigenvalues (more than
+      1e-5 times the maximum real part). If this check fails, it raises a
+      `ValueError`. Otherwise all non-significant imaginary parts that may
```
ok

Commit:
Using double barticks in rst docstring as per review

Commit:
Using double barticks in rst docstring as per review

smarie commented on sklearn/utils/validation.py:
```diff
@@ -966,3 +967,134 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+    It checks:
+
+    - that there are no significant imaginary parts in eigenvalues (more than
+      1e-5 times the maximum real part). If this check fails, it raises a
```
This coefficient, used both to compare imaginary parts and negative parts to the maximum eigenvalue, is purely arbitrary but works in practice. The idea is that numerical approximation errors made by solvers should fall a few orders of magnitude below (1e-7 or more), while a "true" error will fall a few orders of magnitude above (1e-3...). The "gray zone" around the threshold should not in practice being hit so much. We could ask a numerical approximation expert to confirm, if there is one in the core team.

Note that this number could become a parameter so as to make it more obvious that there is a choice being made here, and to allow for future evolutions.

smarie replied:
I now check the dtype at the beginning of the method, thanks for suggesting !

NicolasHug committed:
Update sklearn/utils/validation.py

as per code review

Co-Authored-By: smarie <sylvain.marie@schneider-electric.com>

Commit:
Merge branch 'kPCA_fix_issue_12140' of https://github.com/smarie/scikit-learn into kPCA_fix_issue_12140

# Conflicts:
#	sklearn/utils/validation.py

smarie commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -835,3 +836,63 @@ def test_check_scalar_invalid(x, target_name, target_type, min_val, max_val,
+    input = (5, 4e-12)
+    output = np.array([5., 0.])
+
+    with pytest.warns(None, ) as w:
```
ok

smarie commented on sklearn/utils/validation.py:
```diff
@@ -966,3 +967,134 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+        if warn_on_zeros:
+            warnings.warn("Badly conditioned PSD matrix spectrum: the largest "
+                          "eigenvalue is more than %.2E times the smallest. "
+                          "Small eigenvalues will be replaced with 0"
```
ok

Commit:
End with period (code review)

smarie commented on sklearn/utils/validation.py:
```diff
@@ -966,3 +967,134 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+        if max_imag_abs > 1e-5 * max_real_abs:
+            raise ValueError(
+                "There are significant imaginary parts in eigenvalues (%f "
+                "of the max real part). The matrix is maybe not PSD, or "
```
for max -> ok.

for the percentage symbol:
well, since it is 1e-5  that's 0.00005 so 0.005% . Not sure that the percentage style helps here, but if you think that improves readability I can certainly do it.

NicolasHug committed:
Update sklearn/utils/validation.py

improved readability

Co-Authored-By: smarie <sylvain.marie@schneider-electric.com>

Commit:
max -> maximum (as per review)

Commit:
Merge branch 'kPCA_fix_issue_12140' of https://github.com/smarie/scikit-learn into kPCA_fix_issue_12140

smarie commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -835,3 +836,63 @@ def test_check_scalar_invalid(x, target_name, target_type, min_val, max_val,
+             ('insignificant neg', (5, -5e-5), np.array([5., 0.]), None, ""),]
+
+
+@pytest.mark.parametrize("input, output, w_type, w_msg",
```
ok

smarie commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -835,3 +836,63 @@ def test_check_scalar_invalid(x, target_name, target_type, min_val, max_val,
     assert type(raised_error.value) == type(err_msg)
+
+
+psd_cases = [('nominal', (1., 2.), np.array([1., 2.]), None, ""),
```
That's what I started doing, but it made the code look really ugly - in particular it was much more difficult for maintainers to understand which id relates to which case. So if that has no other consequences I would suggest to leave it as is.

smarie replied:
Note that another style, still involving a variable but even more nice-looking and easy to maintain, would be to use a dict here and use the dict keys as ids and the dict values as cases. I'll probably do that.

smarie commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -835,3 +836,63 @@ def test_check_scalar_invalid(x, target_name, target_type, min_val, max_val,
+                           output)
+
+
+psd_cases2 = [('significant_imag', (5., 5j),
```
see above answer

smarie commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -835,3 +836,63 @@ def test_check_scalar_invalid(x, target_name, target_type, min_val, max_val,
+    with pytest.raises(err_type) as err_info:
+        check_psd_eigenvalues(input)
+
+    assert err_msg in err_info.value.args[0]
```
ok

Commit:
renamed `input` into `lambdas` and `output` into `expected_lambdas`, and transformed cases into dictionaries as per code review

Commit:
using `match` as per review

Commit:
changed comment to match expression

Commit:
Removed explicit floats as per review

Commit:
As per code review: cleaned the kPCA kernel conditioning test as most of it is in the test of `check_psd_eigenvalues`.

smarie commented:
Thanks ! A few conversations still open but only one that is major: the one about numerical thresholds.

Commit:
fixed line length

NicolasHug commented on sklearn/decomposition/kernel_pca.py:
```diff
@@ -79,14 +80,14 @@ class KernelPCA(BaseEstimator, TransformerMixin):
         If RandomState instance, random_state is the random number generator;
         If None, the random number generator is the RandomState instance used
-        by `np.random`. Used when ``eigen_solver`` == 'arpack'.
+        by ``np.random``. Used when ``eigen_solver`` == 'arpack'.
```
Please avoid unrelated changes (I think #13542 will fix this anyway)

When I mentioned double backticks I was only referring to the code you changed ;)

smarie replied:
:)

NicolasHug commented on sklearn/decomposition/tests/test_kernel_pca.py:
```diff
@@ -272,3 +271,18 @@ def test_nested_circles():
+    kpca.fit(X)
+
+    # check that the small non-zero eigenvalue was correctly set to zero
+    assert kpca.lambdas_.min() == 0
```
Maybe also check that `kpca.labmdas_ == check_spd_eig(kpca.labmdas_)`?

smarie replied:
can not do any harm, yes.

Commit:
reverted double barticks fix

Commit:
minor test update as per review: added an assert

Commit:
fixed last test update

Commit:
fixed doctest

NicolasHug commented on doc/whats_new/v0.21.rst:
```diff
@@ -89,6 +89,14 @@ Support for Python 3.4 and below has been officially dropped.
 
+- |Enhancement| :class:`decomposition.KernelPCA` now properly checks the
+  eigenvalues found by the solver for numerical or conditioning issues, relying
+  on new common util `utils.validation.check_psd_eigenvalues`. This ensures
```
``` :func:`utils.validation.check_psd_eigenvalues` ```

smarie replied:
ok

NicolasHug commented on doc/whats_new/v0.21.rst:
```diff
@@ -89,6 +89,14 @@ Support for Python 3.4 and below has been officially dropped.
+- |Enhancement| :class:`decomposition.KernelPCA` now properly checks the
+  eigenvalues found by the solver for numerical or conditioning issues, relying
+  on new common util `utils.validation.check_psd_eigenvalues`. This ensures
+  consistency of results across solvers (different choices for `eigen_solver`),
```
Should be double backticks here, same for ``` `'randomized'` and `'lobpcg'` ```

smarie replied:
ok. I also fixed it for the previous PR text (it seems that it was forgotten on the way)

NicolasHug commented on sklearn/decomposition/kernel_pca.py:
```diff
@@ -84,9 +85,9 @@ class KernelPCA(BaseEstimator, TransformerMixin):
 
     copy_X : boolean, default=True
-        If True, input X is copied and stored by the model in the `X_fit_`
+        If True, input X is copied and stored by the model in the ``X_fit_``
```
Please avoid unrelated changes

smarie replied:
ok

NicolasHug commented on sklearn/decomposition/tests/test_kernel_pca.py:
```diff
@@ -272,3 +273,19 @@ def test_nested_circles():
+
+    # check that the small non-zero eigenvalue was correctly set to zero
+    assert kpca.lambdas_.min() == 0
+    assert_array_equal(kpca.lambdas_, check_psd_eigenvalues(kpca.lambdas_))
```
Use `assert np.allclose()` instead

smarie replied:
why ? it should be exactly the same array (not close, exact)

smarie replied:
ok

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into kPCA_fix_issue_12140

Commit:
added double backticks as per review

Commit:
Removed double backticks in unrelated places

Commit:
using `assert np.all` instead of `assert_array_equal` as per review

Commit:
All thresholds are now variables. Also, added a section about precision.

jnothman commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -835,3 +836,67 @@ def test_check_scalar_invalid(x, target_name, target_type, min_val, max_val,
+    """Test that check_psd_eigenvalues returns the right output for valid
+    input, possibly raising the right warning"""
+
+    with pytest.warns(w_type, match=w_msg) as w:
```
I like the idea of parametrizing this. Nice :)

smarie replied:
:) I got quite involved in writing [pytest plugins](https://github.com/smarie/ALL_OF_THE_ABOVE#tests) to support data science benchmarks creation, so I can't help but using parametrization now :)

jnothman commented on sklearn/utils/validation.py:
```diff
@@ -966,3 +967,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+    Examples
+    --------
+    >>> check_psd_eigenvalues((1, 2))      # nominal case
```
```suggestion
    >>> check_psd_eigenvalues([1, 2])      # nominal case
```
A tuple looks less array-like than a list.

jnothman reviewed (commented):
This looks nice, but I fear my linear algebra is weak. It would be good to hear from someone like @agramfort.

jnothman committed:
Update sklearn/utils/validation.py

Co-Authored-By: smarie <sylvain.marie@schneider-electric.com>

NicolasHug commented on sklearn/utils/validation.py:
```diff
@@ -966,3 +967,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+    # is the provided array in double precision (float64) ?
+    is_double_precision = True
+    try:
+        is_double_precision = lambdas[0].dtype == np.dtype('f')
```
Would that work?

```py
is_double_precision = True
if isinstance(lambdas, np.ndarray):
    is_double_precision = lambdas.dtype == np.dtype('f')
```

You shouldn't need to check `lambdas[0]`

smarie replied:
What is lambdas is a list of numpy elements ? Then the check would be skipped. I admit that this is a twisted case but... can happen no?

NicolasHug replied:
you mean a list of numpy arrays with 1 elements? I don't think it should be supported. 99% of the use case for this function is to take its input from `eigh` which always returns a 1d array

smarie replied:
no I meant a list of numpy elements such as `np.int32(15)`. Elements have a dtype too.
But you're right this is too cumbersome. 

Actually, taking this into consideration, we could actually restrict it to numpy array inputs only. I'll push something like this.

Commit:
Fixed brackets as per review

Commit:
changed double precision check as per review

Commit:
Fixed double precision check as per review, and added corresponding test.

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into kPCA_fix_issue_12140

smarie commented:
> 
> 
> Should this not be checking the width of the float?

indeed ! I was too quick on that one. Fixed and added tests accordingly

Commit:
Fixed line length

Commit:
Fixed unused import

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -981,3 +982,155 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+                "There are significant imaginary parts in eigenvalues (%f "
+                "of the maximum real part). The matrix is maybe not PSD, or "
+                "something went wrong with the eigenvalues decomposition."
+                "" % (max_imag_abs / max_real_abs))
```
```suggestion
                 % (max_imag_abs / max_real_abs))
```

smarie replied:
I personally avoid hanging `%` signs so I always keep the quotes in front. But if that's a style guideline in sklearn I can definitely change

NicolasHug replied:
or just use `format`, but I agree that the empty quote isn't useful

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -981,3 +982,155 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+    # Check that there are no significant imaginary parts
+    if not np.isreal(lambdas).all():
+        max_imag_abs = abs(np.imag(lambdas)).max()
```
```suggestion
        max_imag_abs = np.abs(np.imag(lambdas)).max()
```

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -981,3 +982,155 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+    # Check that there are no significant imaginary parts
+    if not np.isreal(lambdas).all():
+        max_imag_abs = abs(np.imag(lambdas)).max()
+        max_real_abs = abs(np.real(lambdas)).max()
```
```suggestion
        max_real_abs = np.abs(np.real(lambdas)).max()
```

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -981,3 +982,155 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+    Parameters
+    ----------
+    lambdas : array-like
```
```suggestion
    lambdas : array-like, shape (n_eigenvalues,)
```

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -981,3 +982,155 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+    Returns
+    -------
+    lambdas_fixed : array-like
```
```suggestion
    lambdas_fixed : ndarray, shape (n_eigenvalues,)
```

glemaitre reviewed (commented):
LGTM apart of some nitpicks. @agramfort could you have a look at it.

agramfort commented:
looking at this now, I would have kept this new function private
Also I would call the exception PositiveSpectrumWarning to avoid the acronym

glemaitre committed:
Update sklearn/utils/tests/test_validation.py

Co-Authored-By: smarie <sylvain.marie@schneider-electric.com>

glemaitre committed:
Update sklearn/utils/validation.py

Co-Authored-By: smarie <sylvain.marie@schneider-electric.com>

glemaitre committed:
Update sklearn/utils/validation.py

Co-Authored-By: smarie <sylvain.marie@schneider-electric.com>

glemaitre committed:
Update sklearn/utils/validation.py

Co-Authored-By: smarie <sylvain.marie@schneider-electric.com>

glemaitre committed:
Update sklearn/utils/validation.py

Co-Authored-By: smarie <sylvain.marie@schneider-electric.com>

smarie commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -854,3 +855,82 @@ def test_check_scalar_invalid(x, target_name, target_type, min_val, max_val,
+                           output)
+
+
+_psd_cases_invalid = {
```
The link between the ids and the cases is less obvious in this case in my opinion. For 2 cases it is ok but for more (above) it was not so I kept it consistent with the other.

smarie replied:
Closing this conversation if no more activity

smarie commented on doc/whats_new/v0.21.rst:
```diff
@@ -116,6 +116,14 @@ Support for Python 3.4 and below has been officially dropped.
+  ensures consistency of results across solvers (different choices for
+  ``eigen_solver``), including approximate solvers such as ``'randomized'`` and
+  ``'lobpcg'`` (see :issue:`12068`).
+  :issue:`12145` by :user:`Sylvain Marié <smarie>`
```
@NicolasHug suggested 'issue' for pr in another (already merged) pr. Which answer is correct then ?

smarie replied:
Now I see : https://github.com/sloria/sphinx-issues/issues/6

Commit:
`PSDSpectrumWarning` renamed `PositiveSpectrumWarning` as per review

smarie commented:
Thanks @glemaitre , @agramfort . I took into account most of your suggestions. I'll be away for the next 2 weeks, do not hesitate to move on while I'm away though since this is quite stable now and the remaining suggestions are nitpicks.

NicolasHug commented:
I agree it'd be better to keep it private

smarie commented:
Please suggest how to make it private: underscore in name or remove from `__all__` or both ?

Commit:
Hopefully fixed the line length warning

Commit:
Fixed line length

glemaitre commented:
@smarie starting with underscore and removing it form `__all__` as well. In any case, we can import from the module directly even with these constraints.

Commit:
`check_psd_eigenvalues` is now private: renamed to `_check_psd_eigenvalues` and removed from `classes.rst` and `utils/__init__.py`.

smarie commented:
Done, thanks !
 
Ready to merge except if the two open nits' have additional comments from @glemaitre and @NicolasHug .

One test is failing on windows py35 but it does not seem relevant at all to my changes (`test_scale_and_stability`).

jnothman commented:
Yes, it is not relevant to your changes, but it may relate to numerical
conditioning on decomposition :)

smarie commented:
@jnothman apparently not, it happens with the `CCA` model, which has apparently no relation with `KernelPCA`

NicolasHug commented:
the `test_scale_and_stability` failure is loosely related to this PR, in the sense that the failure is caused by some very small eigenvalues that are not (anymore) treated as zero.

@smarie , do you think the new thresholds introduced in https://github.com/scipy/scipy/pull/10067 would make sense here? I like the fact that they directly use `np.finfo`

smarie commented:
Thanks for the explanation and sorry for not clicking faster here :)

If I understand correctly, their strategy is

 - either the user provides a conditioning factor (`cond`) explicitly
 - or in `pinv`, the default `cond = max(a.shape) * np.spacing(a.real.dtype.type(1))`
 - or in `pinv2` and `pinvh`, the default `cond = np.max(np.abs(s)) * max(a.shape) * np.finfo(u.dtype.char.lower()).eps`

We could definitely use the same idea, at least for the "small eigenvalues" (conditioning) detection. As a reminder today we use a hardcoded `cond` with the same value whatever the platform, so it can't hurt. However I do not understand the rationale for the multiplication by `max(a.shape)`. So that means the larger the dimensionality is, the higher the cutoff ? Sounds strange to me but maybe a linear algebra guru would find it reasonable.

Concerning the three other comparisons (significant imaginary part, significant negative value relative/absolute), I have no strong opinion, we can align as well or keep the current code unchanged.
Remember that this private function is currently used only in `KernelPCA`. If its usage expands to more decomposition models (which I hope), we will have the opportunity to think about this again a few times :)

Commit:
Merge remote-tracking branch 'original/master' into kPCA_fix_issue_12140

# Conflicts:
#	sklearn/decomposition/tests/test_kernel_pca.py

smarie committed:
Update sklearn/utils/validation.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

smarie committed:
Update doc/whats_new/v0.21.rst

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

Commit:
Fixed lint issue introduced by code review

smarie commented:
Hi all, this topic is quite old now and seems stable in terms of reviews (many of them, thanks !).

Unless there is strong incentive to align this PR with the new thresholds introduced in scipy/scipy#10067 as suggested by @NicolasHug , I suggest that we adress this topic in another issue/PR rather than here. Indeed the goal of present PR is to be "minimum viable feature" so that #12069 and other similar approximations (e.g. lobpcg) can be applied.

What do you think ?

NicolasHug committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn into pr-12145

NicolasHug committed:
moved whatsnew to 22

NicolasHug committed:
simplify double precision detection

NicolasHug committed:
reduced diff

NicolasHug committed:
update whatsnew

NicolasHug committed:
reduce diff again

NicolasHug reviewed (approved):
Let's merge this, it's been more than a year.

This PR has been stalled because we can't decide whether the proposed thresholds are absolutely perfect, but I guess we'll never know until we try at this point.

Thanks again for the consistent work @smarie 

Maybe @glemaitre @adrinjalali @thomasjpfan want to look at this?

NicolasHug reviewed (dismissed):
Let's merge this, it's been more than a year.

This PR has been stalled because we can't decide whether the proposed thresholds are absolutely perfect, but I guess we'll never know until we try at this point.

Thanks again for the consistent work @smarie 

Maybe @glemaitre @adrinjalali @thomasjpfan want to look at this?

jnothman commented:
+1 for trying to merge this. Happy if a mathematician gives it another look.

adrinjalali commented on sklearn/exceptions.py:
```diff
@@ -171,3 +172,15 @@ class UndefinedMetricWarning(UserWarning):
+    (kernel) present significant negative eigenvalues, or bad conditioning i.e.
+    very small non-zero eigenvalues compared to the largest eigenvalue.
+
+    .. versionadded:: 0.21
```
```suggestion
    .. versionadded:: 0.22
```

adrinjalali commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -937,6 +939,86 @@ def test_check_scalar_invalid(x, target_name, target_type, min_val, max_val,
+            expected_lambdas
+        )
+    if w_type is None:
+        assert not w
```
shouldn't this also check that it's the right warning, if raised?

NicolasHug replied:
It's checked just above `with pytest.warns(w_type, match=w_msg)`

adrinjalali commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -937,6 +939,86 @@ def test_check_scalar_invalid(x, target_name, target_type, min_val, max_val,
+    assert not w
+
+    w_msg = "the largest eigenvalue is more than 1.00E\\+12 times the smallest"
+    with pytest.warns(PositiveSpectrumWarning, match=w_msg) as w:
```
not related to this PR. I think these warnings are exactly the ones which should be on `logging.warn` instead.

smarie replied:
(thanks for the review @adrinjalali  by the way !)

Very good point. see my comment below on the purpose of this. In the subsequent PR requiring this ( #12069 ) I anyway need to disable that warning because it is expected when the approximate decomposition methods for kPCA are used. So the choice is open to you as per this should be using `verbose` + `logging` or `warning`

smarie replied:
EDIT : I got confused between this PR and the subsequent PR. 

So in this PR we need to disable the warning. It is just left as "a possibility for the future".

adrinjalali commented on sklearn/utils/validation.py:
```diff
@@ -1018,6 +1019,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+    It checks:
+
+    - that there are no significant imaginary parts in eigenvalues (more than
+      1e-5 times the maximum real part). If this check fails, it raises a
```
Please update the docstring to reflect the way the threshold is set in the code.

NicolasHug replied:
It's 1e-5 no matter the precision

adrinjalali commented on sklearn/utils/validation.py:
```diff
@@ -1018,6 +1019,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+      remain are removed.
+
+    - that eigenvalues are not all negative. If this check fails, it raises a
+      ``ValueError``
```
Just a note that we have a somewhat similar issue in GPs and there we just raise a warning if the covariance matrix's diagonal (IIRC) has negative values and simply set them to zero.

smarie replied:
indeed that seems to be a similar purpose and similar behaviour (warning + zero-setting). Consistency without sync ! :)

adrinjalali commented on sklearn/utils/validation.py:
```diff
@@ -1018,6 +1019,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+      ``ValueError``
+
+    - that there are no significant negative eigenvalues (with absolute value
+      more than 1e-10 and more than 1e-5 times the largest positive
```
same note about the discrepancy of the code and the docstring here.

adrinjalali commented on sklearn/utils/validation.py:
```diff
@@ -1018,6 +1019,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+    Parameters
+    ----------
+    lambdas : array-like, shape (n_eigenvalues,)
```
```suggestion
    lambdas : array-like of shape (n_eigenvalues,)
```

adrinjalali commented on sklearn/utils/validation.py:
```diff
@@ -1018,6 +1019,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+    lambdas : array-like, shape (n_eigenvalues,)
+        Array of eigenvalues to check / fix.
+
+    warn_on_zeros : bool, optional (default=False)
```
```suggestion
    warn_on_zeros : bool, default=False
```

adrinjalali commented on sklearn/utils/validation.py:
```diff
@@ -1018,6 +1019,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+    Returns
+    -------
+    lambdas_fixed : ndarray, shape (n_eigenvalues,)
```
```suggestion
    lambdas_fixed : ndarray of shape (n_eigenvalues,)
```

adrinjalali commented on sklearn/utils/validation.py:
```diff
@@ -1018,6 +1019,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+    Examples
+    --------
+    >>> _check_psd_eigenvalues([1, 2])      # nominal case
+    ... # doctest: +NORMALIZE_WHITESPACE
```
remove line, this is the default in our configs.

adrinjalali commented on sklearn/utils/validation.py:
```diff
@@ -1018,6 +1019,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([1, 2])
+    >>> _check_psd_eigenvalues([5, 5j])     # significant imag part
+    ... # doctest: +ELLIPSIS +NORMALIZE_WHITESPACE
```
same here

adrinjalali commented on sklearn/utils/validation.py:
```diff
@@ -1018,6 +1019,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+        of the maximum real part). The matrix is maybe not PSD, or something
+        went wrong with the eigenvalues decomposition.
+    >>> _check_psd_eigenvalues([5, 5e-5j])  # insignificant imag part
+    ... # doctest: +NORMALIZE_WHITESPACE
```
here as well, and the rest of them.

adrinjalali commented on sklearn/utils/validation.py:
```diff
@@ -1018,6 +1019,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+    """
+
+    lambdas = np.array(lambdas)
```
I feel like if this check is here, it should be `check_array`

NicolasHug replied:
what check?

adrinjalali replied:
to me, this is checking that the `lambdas` are convertible to a numpy array, and if they are, but are not an array, it converts them to an array. This logic is usually what we do in `check_array`

NicolasHug replied:
We also want to copy the input as well. No strong opinion but `check_array` seems overkill to me since in effect what it will do is precisely call `np.array`

adrinjalali replied:
I think we should avoid using `np.array` in the codebase unless really necessary, so that we can later on handle non-numpy input easier in the future.

NicolasHug replied:
For user input I agree, but in this specific case less so.

In any case, I tried using check_array but it raises an error on complex values.

adrinjalali replied:
If it's not data given by the user, then why do we need to convert to an array in the first place?

NicolasHug replied:
Mostly to make a copy.

It also makes tests easy by providing lists.

adrinjalali replied:
> Mostly to make a copy.

I'd use `lambdas.copy()` if that's what we need to do.

> It also makes tests easy by providing lists.

Ok. I'm not happy about it, but I'll accept it the way it is then.

adrinjalali commented on sklearn/utils/validation.py:
```diff
@@ -1018,6 +1019,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+    if max_eig < 0:
+        raise ValueError("All eigenvalues are negative (maximum is %f). The "
+                         "matrix is maybe not PSD, or something went wrong "
+                         "with the eigenvalues decomposition." % max_eig)
```
I feel like "The matrix is maybe not PSD, or something went wrong with the eigenvalues decomposition." could be written in a better way (and it's repeated a few times here). But not sure how.

NicolasHug replied:
I find it OK but feel free to suggest / edit

adrinjalali replied:
maybe "Either the matrix is not PSD, or there's an issue while computing the eigendecomposition of the matrix".

adrinjalali commented on sklearn/utils/validation.py:
```diff
@@ -1018,6 +1019,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+    lambdas[lambdas < 0] = 0
+
+    # Finally check for conditioning
+    too_small_lambdas = (0 < lambdas) & (lambdas < small_pos_ratio * max_eig)
```
with the current logic, `max_eig` could be negative.

NicolasHug replied:
No, we raise a ValueError above if `max_eig < 0`

adrinjalali commented on sklearn/utils/validation.py:
```diff
@@ -1018,6 +1019,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
         raise ValueError('`{}`= {}, must be <= {}.'.format(name, x, max_val))
 
 
+def _check_psd_eigenvalues(lambdas, warn_on_zeros=False):
```
This warns under several circumstances, I'd probably rather call the parameter `warn` or something.

NicolasHug replied:
I think @smarie had a reason for that, probably related to his follow-up PR https://github.com/scikit-learn/scikit-learn/pull/12069?

adrinjalali replied:
Yes, but it has gone beyond simply checking for small values.

smarie replied:
let me have a look

smarie replied:
`_check_psd_eigenvalues` automatically sets to a proper zero all eigenvalues that are extremely small. I need this feature in PR #12069 , because it relies on an approximate method that very often stops at very small numbers instead of proper zeros.

But then I thought that in some (most?) cases, the caller might be interested in knowing that this operation actually happened. So I added this argument, that is probably not well-named and should rather be renamed `bad_conditioning_warning`. Maybe its default value should be actually `True`, and PR #12069 should explicitly set it to false when using approximate methods.

What do you think ?

adrinjalali replied:
I'd be happy with the renaming to `bad_conditioning_warning`.

If the estimators pass `False` here, would there be any way through the public API that the user sees these warnings?

smarie replied:
For the renaming: @NicolasHug do you agree too ?
For the default value to `True`: do you both agree ?

For your last question @adrinjalali : if an estimator chooses to pass `False`, it purposedly disables the warning because it thinks that having small non-zero eigenvalues is a normal behaviour (due to numerical approximation). So no, there would not then be any way that the user sees this particular warning. The user would still see the other warnings though (significant negative eigenvalue).

NicolasHug replied:
> For the renaming: @NicolasHug do you agree too ?

That means that parameter controls all the warnings, but I'm OK with it

>  For the default value to True: do you both agree ?

I'd leave it to False in this PR (better compatibility with master).
We can change it to True later in  #12069 if we need to.

THanks @smarie

adrinjalali replied:
that means the user can pass a non-PSD matrix and the estimator would do something w/o giving any warnings?

This PR is supposed to fix an issue which says "raise Errors and Warnings according to eigenvalue decomposition numerical/conditioning issues", and we're not doing that if it's always False.

I'm happy with what this PR is doing in terms of fixing those issues in matrices, but we haven't figured how to show those messages to the user. I'd be more comfortable with it being always True for now, until people complain and we find a way to suppress them.

NicolasHug replied:
> and we're not doing that if it's always False.

Yes, we are. We still raise a ValueError in the most problematic cases.

As you said, we're not sure what to do with those warnings. So having the default to False is the safe alternative. It's what master is currently doing. I don't think we should rely on users complaining about it to fix it.

adrinjalali replied:
Ok, then we can merge this with the proposal you have, and open an issue immediately on finding a way to actually show the warnings which are never shown for now.

NicolasHug replied:
sounds good to me

smarie replied:
I think that there was a misunderstanding here

smarie replied:
The `bad_conditioning_warning` option only applies to one warning (the last one) and does not at bypasses the other errors and warnings

NicolasHug replied:
@adrinjalali 's point is that it should control all the warnings, not just one.

Also I think `bad_conditioning_warning` isn't a great name since we also raise errors when the conditionning is bad.

smarie replied:
I updated the PR in that sense, please have a look

smarie replied:
I re-updated the PR again with a very minor update: the comment in the kPCA estimator nos mentions `"make sure that the eigenvalues are ok and fix numerical issues"`.

NicolasHug replied:
@smarie:

- the default should be false
- the parameter should control all the warnings
- there are  still references to `warn_on_zeros` left in the code and docs
- `bad_conditioning_warning` isn't a good name. Maybe `also_warn` would be better.

smarie replied:
Sorry @NicolasHug i skipped your comment, please find my answers here:

> it should control all the warnings, not just one.

There are only 2 kind of warnings in current proposal (the rest are errors) :

 - **significant negative eigenvalues**. This quite problematic for a PSD matrix so I think that the warning should always be displayed. 

 - **small positive eigenvalues**. This is not really a problem, , and it is very frequent in kPCA, since in kernel PCA I think that the matrix often has many trailing very small / zero eigenvalues (this is why the reconstruction using only a small number of eigenvectors is ok, hence the huge perf improvements of approximate decomposition methods). That's why it is really safe to ignore these warnings in kPCA estimator, as it is "nominal" for kPCA. Therefore we added the `bad_conditioning_warning` option to completely ignore the warning.

Now there is no warning currently for **small negative eigenvalues** (they are silently converted to zero without warning in all cases). Should we create one ? If so, should the `bad_conditioning_warning` parameter apply to both small positive and negative eigenvalues ? I think so, but I'd like to hear your thoughts.

NicolasHug replied:
I'll leave you with @adrinjalali to discuss this. I have no strong opinion.

smarie replied:
ok. I'll wait for his reply and discuss with him tomorrow then. Thanks a lot for driving the discussions @NicolasHug !

smarie replied:
@adrinjalali : I updated the proposal according to the following thoughts:

 - since the term "bad conditioning" does not seem to really apply to matrices that have exact zero eigenvalues, I reverted the naming for something more explicit (`small_nonzeros_warning`).

 - I updated the code so that this warning is raised if there are small non-zero **negative or positive** eigenvalues (the warning was raised for positive small previously but not for negative small, which was not consistent)

Let me know if the result is consistent and intuitive enough so that we can mark this as resolved

adrinjalali reviewed (commented):
As far as I can see, we never call the function with `warn_on_zero=True`. We probably want to take `verbose` into account for that or something? And I'm really not comfortable with using warnings for this.

jnothman commented:
Hi @smarie, it would be awesome to have this in our next release (which is coming soon), but @adrinjalali has provided some careful comments. Do you have the chance to update this soon? Thanks

NicolasHug commented:
> Do you have the chance to update this soon? 

If not, no problem at all, we'll take care of it.

jnothman commented:
does this work relate to #8531 (and #8368)?

jnothman commented:
With no reply from @smarie in a few days, should the core devs consider resolving @adrinjalali's comments?

NicolasHug commented:
I'll do it

NicolasHug committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn into pr-12145

NicolasHug committed:
Addressed most comments

NicolasHug commented:
Most comments were addressed @adrinjalali 

>  And I'm really not comfortable with using warnings for this.

why? I think this is common?

adrinjalali commented:
> why? I think this is common?

Yes, and that's an issue. We should at some point start differentiating between a warning raised by `warnings` and simply giving information to the user. To me, this is a good example of information given to the user, and that's why I rather have `looging.warn` and not `warnings.warn`. As I understand, if we raise a warning, we should also give clear directives on how the user can fix the issue. This is not one which can be fixed easily by the user, it's just an information.

smarie commented:
Hi, sorry I was away last week, please let me know if/how I should help.

NicolasHug committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn into pr-12145

NicolasHug committed:
update error message

jnothman commented:
Shall we hold this until after the release?

NicolasHug commented:
I'd rather not that would be the third time we do this. 

I addressed all of @adrinjalali's comments except maybe for https://github.com/scikit-learn/scikit-learn/pull/12145/files#r338919810 which I wouldn't say is critical

smarie commented:
@jnothman  

> does this work relate to #8531 (and #8368)?

No, but I guess #12069 does, since it uses the new `_check_psd_eigenvalues` provided by this PR. So it might fix the issue that #8531 (and #8368) are trying to fix, we'll see.

Commit:
Fixed error message (2) and updated doctests accordingly

Commit:
Renamed `warn_on_zeros` to `bad_conditioning_warning` as per review, and made its default value to `True`.

Commit:
Fixed comment for clarity on why we use `_check_psd_eigenvalues`

Commit:
Minor docstring update

Commit:
`bad_conditioning_warning` renamed to `small_nonzeros_warning` because "bad conditioning" does not really apply when there are also exact zeros.
This warning is now raised if there are small non-zero **negative or positive** eigenvalues (the warning was raised for positive small previously but not for negative small, which was not consistent)

Commit:
fixed linter error

NicolasHug commented on sklearn/utils/validation.py:
```diff
@@ -1020,6 +1021,147 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
         raise ValueError('`{}`= {}, must be <= {}.'.format(name, x, max_val))
 
 
+def _check_psd_eigenvalues(lambdas, small_nonzeros_warning=True):
```
@smarie please leave that default to false

smarie replied:
I will align of course to the majority vote but I do not agree with setting it to `False` here ; since it is already set to `False` at usage time in kernel PCA you should be happy with it, no ? [see here](https://github.com/scikit-learn/scikit-learn/pull/12145/files/13e2ba6de478ce6c3d2b59c84e1ab07eebdc51c1#diff-703280df0f9d49445c95026f6dbbee9fR216)

In the more general case of *other estimators* using the `_check_psd_eigenvalues` utility method, I think that this would be a better default to have it `True`. That way, people not aware of numerical imprecision or bad conditioning will be automaticallly aware about it. Let's not miss a chance to spot errors in the future !

Again, if the above arguments do not convince you I will modify it. Sorry for the over-engineering :)

NicolasHug replied:
We currently only have one estimator using the function. It's using False. There's no need to address *potential* use-cases at the moment.

smarie replied:
Ok then, aligning with you and setting default to `False`.

NicolasHug commented:
@adrinjalali are you ok with:

- having only one parameter to control the warnings
- leaving that false by default
- discussing everything else in other PRs / issues

We are clearly over engineering this PR by trying too hard to future-proof it. Let's keep things simple.
(I would even be OK to remove all the warnings. The ValueErrors are already an improvement).

Commit:
`_check_psd_eigenvalues` : as per code review, `small_nonzeros_warning` is now `False` by default

smarie commented:
a precision concerning your last sentences @NicolasHug  "one parameter to control all the warnings" and "I would be ok to remove all the warnings": please have a look at my previous comments, there is a huge difference between the warning for significant negative values (which could even be transformed into a `ValueError`) and the warning for small non-zeros.

adrinjalali commented:
I'm happy with a significant negative value to be a `ValueError`, but the rest should all be controlled by one single parameter. And for the rest, I'm happy with your proposal @NicolasHug

adrinjalali commented:
Also, @smarie I really appreciate your persistence and you following the suggestions so promptly. I'm sorry it's being dragged for longer than we all would prefer :)

smarie commented:
Perfect. I make the changes right now and push what should finally be the final proposal. Almost there :) !

Commit:
(as per code review)

 * Renamed `small_nonzeros_warning` into `enable_warnings`.
 * now consistent warnings are raised for all three cases (imaginary parts, negative, small non zero), and the parameter disables all of them.
 * improved string formatting using `%g` instead of `%f` or other things

smarie commented:
Ready for a last round. That "simple" last change was actually quite impactant but I think that the result is now straightforward and consistent. I updated the docstring, please have a look.

In details:

 * Renamed `small_nonzeros_warning` into `enable_warnings`.
 * now consistent warnings are raised for all three cases (imaginary parts, negative, small non zero), and the parameter disables all of them.
 * improved string formatting using `%g` instead of `%f` or other things
 * I updated the tests

Commit:
Fixed doctest + One of the error messages was still not consistent with the others. Now adopting the same message everywhere.

Commit:
One test case (small pos eigenvals) had been mistakenly removed and not copied back. Added it.

Commit:
Reverted bad line: `np.real(lambdas)` should also be done if imaginary parts are all zeros, to convert to float dtype

Commit:
Two very minor improvements of the warning messages: now all warning messages explicitly state what is happening (setting to zero the imag/negative/small value).

smarie commented:
All seems ok now, stopping edits and waiting for final review @adrinjalali @NicolasHug .

NicolasHug commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -937,6 +939,81 @@ def test_check_scalar_invalid(x, target_name, target_type, min_val, max_val,
+            _check_psd_eigenvalues(lambdas, enable_warnings=enable_warnings),
+            expected_lambdas
+        )
+    if w_type is None or not enable_warnings:
```
that last part is redundant

```suggestion
    if w_type is None:
```

smarie replied:
indeed

NicolasHug commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -937,6 +939,81 @@ def test_check_scalar_invalid(x, target_name, target_type, min_val, max_val,
+}
+
+
+@pytest.mark.parametrize("lambdas, err_type, err_msg",
```
The `err_type` isn't necessary since they're all ValueError

smarie replied:
good catch. Shall we leave it like that still, for the sake of not reopening the editor on my side ? (and because in the future we might appreciate to have this possibility of creating specific error types?)

NicolasHug replied:
sure

NicolasHug reviewed (approved):
Thanks @smarie, only nitpicks from me but LGTM

smarie committed:
Update sklearn/utils/tests/test_validation.py

Co-Authored-By: Nicolas Hug <contact@nicolas-hug.com>

smarie commented:
Thanks very much for the review. @adrinjalali you're up :)

NicolasHug commented:
ping @adrinjalali

adrinjalali commented on sklearn/decomposition/tests/test_kernel_pca.py:
```diff
@@ -270,3 +271,19 @@ def test_nested_circles():
+
+
+def test_kernel_conditioning():
+    """Test that ``_check_psd_eigenvalues`` is correctly called."""
```
mention `#12140` for future reference

smarie replied:
like this ?

```suggestion
    """ Test that ``_check_psd_eigenvalues`` is correctly called (see 
    :issue:`12040`)."""
```

NicolasHug replied:
yup

NicolasHug replied:
Though it's #12140 not #12040

smarie replied:
thanks, I fixed it

adrinjalali reviewed (approved):
LGTM, thanks a lot @smarie and @NicolasHug

smarie committed:
Update sklearn/decomposition/tests/test_kernel_pca.py

Commit:
Fixed docstring

NicolasHug commented:
Merged!
Thank you so much @smarie for your consistent work on this

jnothman commented:
Yay!!

smarie commented:
So cool ! I will now be able to finalize #12069
Thanks so much for the hard reviewing task !
